### PR TITLE
Supporting connect timeout server option

### PR DESCRIPTION
**Pull request by Shriramu2002**
### Description

In this commit, we add support for the connect timeout server option using the sp_serveroption stored procedure. It specifies the time to wait in seconds before returning from a failed attempt to login to a linked server. 

The connect timeout server option is implemented using the free_tds library API dbsetlogintime().

Task: BABEL-4274
Signed-off-by: Shriramu A R [shriraar@amazon.com](mailto:shriraar@amazon.com)

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** N/A

* **Arbitrary inputs -** Yes

* **Negative test cases -** Yes

* **Minor version upgrade tests -** Added new files to cover minor version  upgrade tests

* **Major version upgrade tests -** Already covered using the existing openquery files

* **Performance tests -** N/A

* **Tooling impact -** N/A

* **Client tests -** N/A

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/linked_servers.c [hunk 1]
```diff
@@ -742,7 +742,8 @@ linked_server_establish_connection(char *servername, LinkedServerProcess * lspro
 	ListCell	*option;
 	char	*data_src = NULL;
 	char	*database = NULL;
-	int 	query_timeout = 0;
+	int		query_timeout = 0;
+	int		connect_timeout = 0;
```
Comment (basasairohan) [inline contrib/babelfishpg_tsql/src/linked_servers.c]: indentation

File: contrib/babelfishpg_tsql/src/catalog.c [hunk 2]
```diff
@@ -1237,14 +1237,14 @@ get_bbf_servers_def_idx_oid()
 	return bbf_servers_def_idx_oid;
 }
 
-int
-get_query_timeout_from_server_name(char *servername)
+int 
+get_timeout_from_server_name(char *servername,int attnum)
```
Comment (basasairohan) [inline contrib/babelfishpg_tsql/src/catalog.c]: add a space after the comma.

File: contrib/babelfishpg_tsql/src/linked_servers.c [hunk 3]
```diff
@@ -801,12 +802,19 @@ linked_server_establish_connection(char *servername, LinkedServerProcess * lspro
 						 ));
 		}
 
+		/* fetch query timeout from the servername */
+		query_timeout = get_timeout_from_server_name(servername, Anum_bbf_servers_def_query_timeout);
+
+		/* fetch query timeout from the servername */
```
Comment (basasairohan) [inline contrib/babelfishpg_tsql/src/linked_servers.c]: connect timeout*

File: contrib/babelfishpg_tsql/src/linked_servers.c [hunk 4]
```diff
@@ -801,12 +802,19 @@ linked_server_establish_connection(char *servername, LinkedServerProcess * lspro
 						 ));
 		}
 
+		/* fetch query timeout from the servername */
+		query_timeout = get_timeout_from_server_name(servername, Anum_bbf_servers_def_query_timeout);
```
Comment (basasairohan) [inline contrib/babelfishpg_tsql/src/linked_servers.c]: why are we fetching query timeout here? 
We are fetching it below when isTesting is false, right?

File: contrib/babelfishpg_tsql/src/linked_servers.h [hunk 5]
```diff
@@ -102,6 +102,7 @@ typedef DBPROCESS * LinkedServerProcess;
 #define LINKED_SERVER_SET_VERSION(login)		DBSETLVERSION(login, DBVERSION_74)
 #define LINKED_SERVER_SET_DBNAME(login, dbname)		DBSETLDBNAME(login, dbname)
 #define LINKED_SERVER_SET_QUERY_TIMEOUT(timeout) 	dbsettime(timeout)
+#define LINKED_SERVER_SET_CONNECT_TIMEOUT(timeout) dbsetlogintime(timeout)
```
Comment (basasairohan) [inline contrib/babelfishpg_tsql/src/linked_servers.h:105]: it should be ((void)0) if ENABLE_TDS_LIB flag is not used.

File: contrib/babelfishpg_tsql/src/plerrcodes.h [hunk 6]
```diff
@@ -1007,3 +1007,4 @@
 {
 	"pltsql_error_not_mapped", ERRCODE_PLTSQL_ERROR_NOT_MAPPED
 },
+
```
Comment (basasairohan) [inline contrib/babelfishpg_tsql/src/plerrcodes.h]: ?

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 7]
```diff
@@ -2688,12 +2722,16 @@ sp_serveroption_internal(PG_FUNCTION_ARGS)
 	while (*newoptionvalue != '\0' && isspace((unsigned char) *newoptionvalue))
 		newoptionvalue++;
 
-	if (optionname && strlen(optionname) == 13 && strncmp(optionname, "query timeout", 13) == 0)
-		update_bbf_server_options(servername, optionname, newoptionvalue, false);
+	if (optionname && (strlen(optionname) == 13 || strlen(optionname) == 15) && (strncmp(optionname, "query timeout", 13) == 0 || strncmp(optionname, "connect timeout", 15) == 0))
```
Comment (basasairohan) [inline contrib/babelfishpg_tsql/src/procedures.c]: what if optionname = "query timeoutxy" ?

File: contrib/babelfishpg_tsql/src/linked_servers.c [hunk 8]
```diff
@@ -801,12 +802,15 @@ linked_server_establish_connection(char *servername, LinkedServerProcess * lspro
 						 ));
 		}
 
+		/* fetch query timeout from the servername */
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/linked_servers.c]: comment should say connect timeout

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 9]
```diff
@@ -2437,41 +2437,68 @@ update_bbf_server_options(char *servername, char *optname, char *optvalue, bool
 	ScanKeyData		key;
 	HeapTuple		tuple, old_tuple;
 	TableScanDesc	tblscan;
+	int		nargs = isInsert?BBF_SERVERS_DEF_NUM_COLS - 1:1;
 
-	if (optname != NULL && strlen(optname) == 13 && strncmp(optname, "query timeout", 13) == 0)
+	if(optname == NULL || optvalue == NULL)
+		{
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c]: Fix indentation

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 10]
```diff
@@ -2427,7 +2427,7 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
 }
 
 static void 
-update_bbf_server_options(char *servername, char *optname, char *optvalue, bool isInsert)
+update_bbf_server_options(char *servername, char **optname, char **optvalue, bool isInsert)
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c]: The hardcoded arrays can be defined globally (since these won't change) and we can use those every time an insert happens. We can ignore the optname and optvalue arguments and just do a catalog insert using the default values.

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 11]
```diff
@@ -2875,6 +2907,8 @@ sp_serveroption_internal(PG_FUNCTION_ARGS)
 	char *optionname = PG_ARGISNULL(1) ? NULL : lowerstr(text_to_cstring(PG_GETARG_VARCHAR_PP(1)));
 	char *optionvalue = PG_ARGISNULL(2) ? NULL : lowerstr(text_to_cstring(PG_GETARG_VARCHAR_PP(2)));
 	char *newoptionvalue = optionvalue;
+	char *optname[1];
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c]: why are these lines needed?

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 12]
```diff
@@ -2553,6 +2583,8 @@ sp_addlinkedserver_internal(PG_FUNCTION_ARGS)
 	char	   *data_src = PG_ARGISNULL(3) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(3));
 	char	   *provstr = PG_ARGISNULL(5) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(5));
 	char	   *catalog = PG_ARGISNULL(6) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(6));
+	char	   * optname[BBF_SERVERS_DEF_NUM_COLS - 1] = {"query timeout", "connect timeout"};
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c]: these are only to be defined once so we can make them global

File: contrib/babelfishpg_tsql/src/linked_servers.c [hunk 13]
```diff
@@ -851,6 +860,7 @@ linked_server_establish_connection(char *servername, LinkedServerProcess * lspro
 			LINKED_SERVER_SET_QUERY_TIMEOUT(query_timeout);
 		}
 
+
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/linked_servers.c]: nit: unnecessary newline

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 14]
```diff
@@ -99,6 +99,9 @@ char	   *sp_describe_first_result_set_view_name = NULL;
 bool		sp_describe_first_result_set_inprogress = false;
 char	   *orig_proc_funcname = NULL;
 
+char	   * srvOptions_optname[BBF_SERVERS_DEF_NUM_COLS - 1] = {"query timeout", "connect timeout"};
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c:103]: Add a comment saying what these mean

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 15]
```diff
@@ -2437,84 +2440,122 @@ update_bbf_server_options(char *servername, char *optname, char *optvalue, bool
 	ScanKeyData		key;
 	HeapTuple		tuple, old_tuple;
 	TableScanDesc	tblscan;
+	int		nargs = BBF_SERVERS_DEF_NUM_COLS - 1;
+
+	MemSet(new_record_repl, false, sizeof(new_record_repl));
 
-	if (optname != NULL && strlen(optname) == 13 && strncmp(optname, "query timeout", 13) == 0)
+	if(isInsert && optname == NULL && optvalue == NULL)
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c]: checking for optname and optvalue is redundant, we can just skip it since we anyways will not use it in case of an insert

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 16]
```diff
@@ -2437,84 +2440,122 @@ update_bbf_server_options(char *servername, char *optname, char *optvalue, bool
...
+	{
+		for(int i = 0; i < nargs; i++)
+		{
+			if((strlen(srvOptions_optname[i]) == 13 && strncmp(srvOptions_optname[i], "query timeout", 13) == 0) || (strlen(srvOptions_optname[i]) == 15 && strncmp(srvOptions_optname[i], "connect timeout", 15) == 0))
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c:2454]: add a comment saying that this check is needed for options where we wish to convert string input to int

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 17]
```diff
@@ -2437,84 +2440,122 @@ update_bbf_server_options(char *servername, char *optname, char *optvalue, bool
...
+			{
+				int32	timeout;
+				timeout = atoi(srvOptions_optvalue[i]);
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c]: This can be one line

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 18]
```diff
@@ -2437,84 +2440,122 @@ update_bbf_server_options(char *servername, char *optname, char *optvalue, bool
...
+					new_record[Anum_bbf_servers_def_query_timeout - 1] = Int32GetDatum(timeout);
+				else
+					new_record[Anum_bbf_servers_def_connect_timeout - 1] = Int32GetDatum(timeout);
+			}
+		}
+	}
+	else
 	{
-		int32		query_timeout;
+		if(optname!= NULL && (strlen(optname) == 13 || strlen(optname) == 15) && (strncmp(optname, "query timeout", 13) == 0 || strncmp(optname, "connect timeout", 15) == 0))
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c]: rearrange this condition. if option name is "query timeoux" then we will do strncmp(optname, "connect timeout", 15) which is not something we want. Make this

`optname != NULL && ((strlen(optname) == 13 && strncmp(optname, "query timeout", 13) == 0) || (strlen(optname) == 15 && strncmp(optname, "connect timeout", 15) == 0))`

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 19]
```diff
@@ -2437,84 +2440,122 @@ update_bbf_server_options(char *servername, char *optname, char *optvalue, bool
...
+			if(!new_record_repl[i])
 			{
-				table_endscan(tblscan);
-				table_close(bbf_servers_def_rel, RowExclusiveLock);
-				ereport(ERROR,
-						(errcode(ERRCODE_FDW_ERROR),
-				 		errmsg("The server '%s' does not exist. Use sp_linkedservers to show available servers.", servername)));
+				bool isNull;
+				new_record[i] = heap_getattr(old_tuple, i+1,
+												RelationGetDescr(bbf_servers_def_rel), &isNull);
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c:2544]: we need not fetch the attributes again, this adds unnecessary overhead. We can simply set the new_record[] like we are doing for the servername

File: test/JDBC/expected/openquery_upgrd_before_15_4-vu-prepare.out [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+
+-- sp_serveroption with valid server, optname and optvalue
+EXEC sp_serveroption @server='bbf_server_1', @optname='query timeout', @optvalue='1'
+GO
+
+EXEC sp_serveroption @server='bbf_server_2', @optname='connect timeout', @optvalue='1'
+GO
+
+-- check if the linked servers added above are reflected in babelfish_server_options catalog
+SELECT servername, query_timeout, connect_timeout FROM babelfish_server_options WHERE servername = 'bbf_server' OR servername = 'bbf_server_1'
```
Comment (thephantomthief) [inline test/JDBC/expected/openquery_upgrd_before_15_4-vu-prepare.out]: add bbf_server_2 in WHERE clause as well

File: test/JDBC/expected/openquery-vu-prepare.out [hunk 21]
```diff
@@ -228,17 +339,18 @@ varchar#!#varchar#!#varchar#!#jsonb
...
-SELECT servername, query_timeout FROM babelfish_server_options WHERE servername = 'bbf_server' OR servername = 'bbf_server_1'
+SELECT servername, query_timeout, connect_timeout FROM babelfish_server_options WHERE servername = 'bbf_server' OR servername = 'bbf_server_1'
 GO
 ~~START~~
-varchar#!#int
-bbf_server#!#0
-bbf_server_1#!#1
+varchar#!#int#!#int
+bbf_server#!#0#!#0
+bbf_server_1#!#1#!#1
```
Comment (thephantomthief) [inline test/JDBC/expected/openquery-vu-prepare.out]: We can remove this entirely as discussed with @basasairohan

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 22]
```diff
@@ -2553,7 +2596,6 @@ sp_addlinkedserver_internal(PG_FUNCTION_ARGS)
 	char	   *data_src = PG_ARGISNULL(3) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(3));
 	char	   *provstr = PG_ARGISNULL(5) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(5));
 	char	   *catalog = PG_ARGISNULL(6) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(6));
-
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c]: Undo this change

**Review by thephantomthief — APPROVED** [review]

**Review by basasairohan — APPROVED** [review]

**Review by thephantomthief — APPROVED** [review]

**Review by shardgupta — APPROVED** [review]
